Initial commit of JmxScraper code.

robsunday opened the pull request:
In scope:

- Application compiles and runs as standalone and can read config file.
- No JMX connections yet.
- Typo in project name corrected

In few places I added temporary @SuppressWarnings. They will be removed after later on.
I will get rid of prometheus HTTP endpoint related code in next PR

renovate[bot] committed:
Update dependency gradle to v8.10.1 (#1447)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update micrometer packages to v1.13.4 (#1448)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update dependency com.linecorp.armeria:armeria-bom to v1.30.1 (#1449)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

robsunday committed:
Initial commit of JmxScraper code. Application compiles and runs as standalone and can read config file.

renovate[bot] committed:
Update plugin com.gradle.develocity to v3.18.1 (#1451)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update dependency com.gradle.enterprise:com.gradle.enterprise.gradle.plugin to v3.18.1 (#1450)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

robsunday committed:
Cleanup of config.
Refactoring of JmxScraperConfigFactory.
Added first unit tests.

SylvainJuge commented on jmx-scraper/src/main/java/io/opentelemetry/contrib/jmxscraper/JmxScraper.java:
```diff
@@ -0,0 +1,78 @@
+
+  private void shutdown() {
+    logger.info("Shutting down JmxScraper and exporting final metrics.");
+    exec.shutdown();
```
[minor] even if there is only one task executed, making sure that the pool is properly shut down with code similar to https://github.com/elastic/apm-agent-java/blob/48d2a231e604414ccb4385792a8c5e364567dcd9/apm-agent-core/src/main/java/co/elastic/apm/agent/util/ExecutorUtils.java#L239

robsunday replied:
Thanks, great suggestion!

SylvainJuge commented on jmx-scraper/src/main/java/io/opentelemetry/contrib/jmxscraper/config/JmxScraperConfigFactory.java:
```diff
@@ -0,0 +1,230 @@
+import java.util.stream.Collectors;
+
+@SuppressWarnings({"SystemOut", "SystemExitOutsideMain"})
+public class JmxScraperConfigFactory {
```
[minor] this factory could have been moved to a static method factory in JmxScraperConfig to avoid having a separate class just for parsing it.

robsunday replied:
So in the original code from jmx-metrics it was done the way you suggest. Config class contained a lot of logic related to parsing and just some data fields.
I intentionally extracted parsing logic to separate class so config is only responsible for carrying the data, and factory is responsible creating this object, since the task is not trivial.
I also like current implementation also because in the past Config had a lot of constants needed only for parsing and system properties processing. Now it looks cleaner to me, but I'm open to discussion.

SylvainJuge replied:
Ok, then let's keep it as-is for now, we can always revisit this later.

SylvainJuge commented on jmx-scraper/src/main/java/io/opentelemetry/contrib/jmxscraper/config/JmxScraperConfigFactory.java:
```diff
@@ -0,0 +1,230 @@
+  public JmxScraperConfig createConfigFromArgs(List<String> args) {
+    if (!args.isEmpty() && (args.size() != 2 || !args.get(0).equalsIgnoreCase("-config"))) {
+      System.out.println(
+          "Usage: java io.opentelemetry.contrib.jmxscraper.JmxScraper "
```
The actual command is based on `java -jar` as the main class is provided in the manifest. If we really want to have the original JVM command line we can get it through JMX, but I don't think it's worth the extra complexity.

robsunday replied:
Good point. I blindly copied this code from jmx-metrics. 
Also - the factory is not the place where commandline arguments should be parsed. I'll refactor it.

SylvainJuge commented on jmx-scraper/src/main/java/io/opentelemetry/contrib/jmxscraper/config/JmxScraperConfigFactory.java:
```diff
@@ -0,0 +1,230 @@
+      props.load(is);
+    } catch (IOException e) {
+      System.out.println("Failed to read config properties from stdin: " + e.getMessage());
+      System.exit(1);
```
I think it's better to keep the calls to `System.exit` in `main` and propagate `IOException` in the method signature (or wrap it into a `ConfigurationException`), otherwise that makes this code impossible to unit test.

SylvainJuge commented on jmx-scraper/src/main/java/io/opentelemetry/contrib/jmxscraper/jmx/JmxClient.java:
```diff
@@ -0,0 +1,110 @@
+
+  public JmxClient(JmxScraperConfig config) throws MalformedURLException {
+    this.url = new JMXServiceURL(config.getServiceUrl());
+    ;
```
[minor] leftover `;`

robsunday replied:
Will be removed

SylvainJuge commented on jmx-scraper/src/main/java/io/opentelemetry/contrib/jmxscraper/config/JmxScraperConfig.java:
```diff
@@ -0,0 +1,72 @@
+
+/** This class keeps application settings */
+public class JmxScraperConfig {
+  String serviceUrl = "";
```
[minor] having a static factory allows to make those fields private instead of package private.

robsunday replied:
True, however I still prefer having separate factory. Maybe making the JmxScraperConfigFactory an inner class of JmxScraperConfig would be a good compromise? What do you think? Unfortunately it will make JmxScraperConfig.java quite a big file, which I don't really like...

SylvainJuge replied:
For now let's keep the factory class.

SylvainJuge commented on jmx-scraper/src/test/java/io/opentelemetry/contrib/jmxscraper/config/JmxScraperConfigFactoryTest.java:
```diff
@@ -0,0 +1,325 @@
+
+    // Then
+    assertThat(config.serviceUrl).isEqualTo("jservice:jmx:rmi:///jndi/rmi://localhost:9010/jmxrmi");
+    assertThat(config.customJmxScrapingConfigPath).isEqualTo("");
+    assertThat(config.targetSystems).isEqualTo(new HashSet<>(Arrays.asList("tomcat", "activemq")));
```
```suggestion
    assertThat(config.customJmxScrapingConfigPath).isEmpty();
    assertThat(config.targetSystems).containsOnly("tomcat", "activemq");
```

SylvainJuge commented on jmx-scraper/src/test/java/io/opentelemetry/contrib/jmxscraper/config/JmxScraperConfigFactoryTest.java:
```diff
@@ -0,0 +1,325 @@
+        .hasMessage("Failed to parse otel.jmx.interval.milliseconds");
+  }
+
+  //  @ClearSystemProperty(key = "otel.metric.export.interval")
+
+  //  @Test
+  //  @SetSystemProperty(key = "otel.jmx.service.url", value = "myServiceUrl")
+  //  @SetSystemProperty(key = "otel.jmx.groovy.script", value = "myGroovyScript")
+  //  @SetSystemProperty(
+  //      key = "otel.jmx.target.system",
+  //      value = "mytargetsystem,mytargetsystem,myothertargetsystem,myadditionaltargetsystem")
```
Is it relevant to merge those commented lines right now or are they just a reminder of the extra tests that are left to implement ?

robsunday replied:
Work on this test suite is not finished yet. I wanted to keep it until I'm done. Final code will not heave it.

renovate[bot] committed:
Update dependency io.opentelemetry.instrumentation:opentelemetry-instrumentation-bom-alpha to v2.8.0-alpha (#1456)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: Jack Berg <jberg@newrelic.com>

renovate[bot] committed:
Update errorProneVersion to v2.32.0 (#1453)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
Update plugin com.squareup.wire to v5.1.0 (#1452)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

PeterF778 committed:
Composite Samplers prototype (#1443)

Co-authored-by: Otmar Ertl <otmar.ertl@gmail.com>

psx95 committed:
Fix incorrect cloud.platform value for GCF (#1454)

jefchien committed:
Fix Tomcat metric definitions to aggregate multiple MBeans. (#1366)

yurybubnov committed:
Issue-1034 Short XRay Trace (#1036)

SylvainJuge committed:
add span stacktrace config option (#1414)

Co-authored-by: jackshirazi <jacks@fasterj.com>

jack-berg committed:
Add declarative config support for aws xray propagators (#1442)

brunobat committed:
Fix native mode error cause by static init of random (#862)

renovate[bot] committed:
Update dependency com.uber.nullaway:nullaway to v0.11.3 (#1457)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

jack-berg committed:
Add declarative config support for RuleBasedRoutingSampler (#1440)

robsunday committed:
Code review changes:
Argument parsing moved to the main class.
Argument validation now throw exception.
More tests added for config factory.
Created unit tests for JmxScraper class.

SylvainJuge commented on jmx-scraper/src/main/java/io/opentelemetry/contrib/jmxscraper/jmx/JmxClient.java:
```diff
@@ -0,0 +1,109 @@
+import javax.management.remote.JMXServiceURL;
+
+@SuppressWarnings("unused") // TODO: Temporary
+public class JmxClient {
```
This will need to be replaced with the `RemoteJmxClient` class introduced in https://github.com/SylvainJuge/opentelemetry-java-contrib/pull/2

SylvainJuge commented on jmx-metrics/README.md:
```diff
@@ -35,6 +35,10 @@ Kafka metric-gathering scripts determined by the comma-separated list in `otel.j
 
+Some metrics (e.g. `tomcat.sessions`) are configured to query multiple MBeans. By default, only the value in the first MBean
+is recorded for the metric and all other values are dropped. To aggregate the MBean values together, set the
+`otel.jmx.aggregate.across.mbeans` property to `true`.
```
This is something that we will have to add support in the instrumentation side as well.
